File: internal/tui/tooldetail_antigravity_test.go
package tui

import (
	"strings"
	"testing"

	"github.com/MunifTanjim/argus/internal/transcript"
)

func agyItem(name, input, result string) transcript.Item {
	return transcript.Item{Kind: transcript.ItemTool, ToolName: name, ToolInput: input, Result: result}
}

func assertContains(t *testing.T, out string, wants ...string) {
	t.Helper()
	for _, w := range wants {
		if !strings.Contains(out, w) {
			t.Errorf("missing %q:\n%s", w, out)
		}
	}
}

func assertNotContains(t *testing.T, out string, unwanted ...string) {
	t.Helper()
	for _, u := range unwanted {
		if strings.Contains(out, u) {
			t.Errorf("should not contain %q:\n%s", u, out)
		}
	}
}

func TestGrepSearchDetail(t *testing.T) {
	m := testModel()
	res := `Created At: 2026-07-04T22:12:53+06:00
Completed At: 2026-07-04T22:12:53+06:00
{"File":"/a/b.go","LineNumber":41,"LineContent":"  probe against a"}
{"File":"/a/c.go","LineNumber":9,"LineContent":"probe here"}`
	out := m.toolBody(agyItem("grep_search",
		`{"Query":"probe","SearchPath":"/a","IsRegex":false,"CaseInsensitive":true}`, res), 80)
	assertContains(t, out, `"probe"`, "in /a", "case-insensitive", "/a/b.go:41", "probe against a", "/a/c.go:9")
	assertNotContains(t, out, "Created At", "LineContent")
}

func TestListDirDetail(t *testing.T) {
	m := testModel()
	res := `Created At: x
{"name":".git","isDir":true}
{"name":"Makefile","sizeBytes":"2365"}`
	out := m.toolBody(agyItem("list_dir", `{"DirectoryPath":"/repo"}`, res), 80)
	assertContains(t, out, "/repo", ".git/", "Makefile", "2.3k")
	assertNotContains(t, out, "Created At", "sizeBytes")
}

func TestViewFileDetail(t *testing.T) {
	m := testModel()
	res := "Created At: x\nCompleted At: x\nFile Path: `file:///a.go`\nTotal Lines: 21\nTotal Bytes: 17924\nShowing lines 1 to 21\nThe following code has been modified to include a line number before every line.\n1: package main\n2: func main() {}"
	out := m.toolBody(agyItem("view_file", `{"AbsolutePath":"/a.go"}`, res), 80)
	assertContains(t, out, "/a.go", "Total Lines: 21", "1: package main", "2: func main() {}")
	assertNotContains(t, out, "Created At", "File Path:", "The following code")
}

func TestWriteToFileDetail(t *testing.T) {
	m := testModel()
	res := "Created At: x\nCompleted At: x\nCreated file file:///a.md with requested content.\nIf relevant, proactively run terminal commands to execute this code for the USER. Don't ask for permission."
	out := m.toolBody(agyItem("write_to_file",
		`{"TargetFile":"/a.md","Description":"make notes","CodeContent":"# Title\nbody","Overwrite":true}`, res), 80)
	assertContains(t, out, "make notes", "/a.md", "(overwrite)", "+ # Title", "+ body", "Created file")
	assertNotContains(t, out, "Created At", "proactively run terminal commands")
}

func TestReplaceFileContentDetail(t *testing.T) {
	m := testModel()
	out := m.toolBody(agyItem("replace_file_content",
		`{"TargetFile":"/a.md","Description":"edit","StartLine":6,"EndLine":11,"TargetContent":"- Item 1\n- Item 2","ReplacementContent":"- New 1\n- Item 2"}`, "Created At: x\ndiff echo"), 80)
	assertContains(t, out, "edit", "/a.md", "lines 6", "- - Item 1", "+ - New 1", "- Item 2")
	// result is just the diff echo; shown only on error
	assertNotContains(t, out, "diff echo")
}

func TestMultiReplaceFileContentDetail(t *testing.T) {
	m := testModel()
	out := m.toolBody(agyItem("multi_replace_file_content",
		`{"TargetFile":"/a.md","Description":"multi","ReplacementChunks":[{"StartLine":3,"EndLine":4,"TargetContent":"old a","ReplacementContent":"new a"},{"StartLine":13,"EndLine":14,"TargetContent":"old b","ReplacementContent":"new b"}]}`, ""), 80)
	assertContains(t, out, "/a.md", "(2 edits)", "edit 1", "edit 2", "- old a", "+ new a", "- old b", "+ new b")
}

func TestSearchWebDetail(t *testing.T) {
	m := testModel()
	res := `Created At: x
Completed At: x
The search for "q" returned the following summary:
The **answer** is here.`
	out := m.toolBody(agyItem("search_web", `{"query":"q","domain":"example.com"}`, res), 80)
	assertContains(t, out, `"q"`, "example.com", "answer")
	assertNotContains(t, out, "Created At", "returned the following summary")
}

func TestGenerateImageDetail(t *testing.T) {
	m := testModel()
	res := "Created At: x\nCompleted At: x\nUsing prompt: a logo\n\nGenerated image is saved at /out.jpg.\n\n Do not output the path of this image to show to the user since the user can already see it. However, you can embed this image in artifacts for the USER's review."
	out := m.toolBody(agyItem("generate_image", `{"ImageName":"logo","AspectRatio":"1:1","Prompt":"a logo"}`, res), 80)
	assertContains(t, out, "logo", "1:1", "a logo", "Generated image is saved at /out.jpg")
	assertNotContains(t, out, "Created At", "Using prompt:", "Do not output the path")
}

func TestDefineSubagentDetail(t *testing.T) {
	m := testModel()
	out := m.toolBody(agyItem("define_subagent",
		`{"name":"helper","description":"a helper","system_prompt":"You are a helper.","enable_write_tools":true,"enable_mcp_tools":false,"enable_subagent_tools":false}`, "Created At: x\nSubagent defined."), 80)
	assertContains(t, out, "helper", "a helper", "tools: write", "System Prompt", "You are a helper.")
}

func TestManageTaskDetail(t *testing.T) {
	m := testModel()
	res := "Created At: x\nCompleted At: x\nTask: t-1\nStatus: RUNNING\nLog: /l.log\nLast progress: never\n\nREMINDER: Do not call this tool again to poll."
	out := m.toolBody(agyItem("manage_task", `{"Action":"status","TaskId":"t-1"}`, res), 80)
	assertContains(t, out, "status", "t-1", "Status:", "RUNNING", "Last progress:")
	assertNotContains(t, out, "Created At", "REMINDER")
}

func TestAskQuestionDetail(t *testing.T) {
	m := testModel()
	in := `{"questions":[{"is_multi_select":false,"question":"Pick one","options":["Run all, including a, b","Only files"]}]}`
	res := "Created At: x\nCompleted At: x\nA1: Run all, including a, b"
	out := m.toolBody(agyItem("ask_question", in, res), 80)
	assertContains(t, out, "Agent is asking", "Pick one", "Run all, including a, b", "Only files", "◉")
}

func TestAskPermissionDetail(t *testing.T) {
	m := testModel()
	res := "Created At: x\nCompleted At: x\nPermission for command(echo) was granted. Reason provided by agent: testing."
	out := m.toolBody(agyItem("ask_permission",
		`{"Action":"command","Target":"echo","Reason":"testing"}`, res), 80)
	assertContains(t, out, "command(echo)", "testing", "was granted")
	assertNotContains(t, out, "Created At")
}

func TestSendMessageDetail(t *testing.T) {
	m := testModel()
	out := m.toolBody(agyItem("send_message", `{"Recipient":"agent-1","Message":"hello there"}`, "Created At: x\nMessage sent."), 80)
	assertContains(t, out, "agent-1", "hello there")
}

func TestScheduleDetail(t *testing.T) {
	m := testModel()
	res := "Created At: x\nTool is running as a background task with task id: t-39\nTask Description: Timer: 5s"
	out := m.toolBody(agyItem("schedule", `{"DurationSeconds":"5","Prompt":"ping","TimerCondition":"never"}`, res), 80)
	assertContains(t, out, "Timer 5s", "condition: never", "ping", "t-39")
	assertNotContains(t, out, "Created At")
}

func TestListPermissionsDetail(t *testing.T) {
	m := testModel()
	res := "Created At: x\nCompleted At: x\nYou have read and write access to:\n- /repo"
	out := m.toolBody(agyItem("list_permissions", `{}`, res), 80)
	assertContains(t, out, "read and write access", "/repo")
	assertNotContains(t, out, "Created At")
}

File: app/test/ui/tool_detail_antigravity_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:argus/models/chunk.dart';
import 'package:argus/ui/tool_detail.dart';
import 'package:argus/ui/tool_detail_antigravity.dart';

Widget _wrap(Item i) => MaterialApp(
    home: Scaffold(body: SingleChildScrollView(child: toolDetailBody(i))));

void main() {
  group('result helpers', () {
    test('agyResultBody strips leading timestamp lines', () {
      const r = 'Created At: x\nCompleted At: y\nreal payload\nmore';
      expect(agyResultBody(r), 'real payload\nmore');
    });

    test('agyResultBody leaves untimestamped bodies untouched', () {
      expect(agyResultBody('just text'), 'just text');
    });

    test('stripAgyReminder cuts at REMINDER:', () {
      expect(stripAgyReminder('status: ok\nREMINDER: do not poll'), 'status: ok');
    });

    test('stripAgyBoilerplate removes the appended instruction', () {
      const r =
          "Created file.\nIf relevant, proactively run terminal commands to execute this code for the USER. Don't ask for permission.";
      expect(stripAgyBoilerplate(r), 'Created file.');
    });

    test('formatBytes scales units', () {
      expect(formatBytes(512), '512B');
      expect(formatBytes(1536), '1.5k');
      expect(formatBytes(1572864), '1.5M');
    });

    test('splitRunCommandResult splits at the tab-indented Output marker', () {
      const r = 'Created At: x\n\tExit Code: 0\n\tOutput:\n\thello\n\tworld';
      final body = agyResultBody(r);
      final s = splitRunCommandResult(body);
      expect(s.hasMarker, isTrue);
      expect(s.head.contains('Exit Code: 0'), isTrue);
      expect(s.output, 'hello\nworld');
    });

    test('splitRunCommandResult with no marker keeps whole body', () {
      final s = splitRunCommandResult('Task started in background.');
      expect(s.hasMarker, isFalse);
      expect(s.output, '');
    });

    test('grepRows formats JSON matches as file:line content', () {
      const body =
          '{"File":"a.dart","LineNumber":12,"LineContent":"  final x = 1;"}';
      expect(grepRows(body), ['a.dart:12 final x = 1;']);
    });

    test('listDirRows marks dirs and sizes files', () {
      const body =
          '{"name":"src","isDir":true}\n{"name":"main.dart","isDir":false,"sizeBytes":"2048"}';
      expect(listDirRows(body), ['src/', 'main.dart  2.0k']);
    });

    test('splitViewFileResult separates meta from numbered content', () {
      const r =
          'Created At: x\nFile Path: `a.go`\nTotal Lines: 2\nTotal Bytes: 20\nShowing lines 1 to 2\nThe following code has line numbers.\n1: package main\n2: func main() {}';
      final v = splitViewFileResult(r);
      expect(v.meta.contains('Total Lines: 2'), isTrue);
      expect(v.content, '1: package main\n2: func main() {}');
    });

    test('parseAgyAnswers maps A1/A2 to 0-based indices', () {
      expect(parseAgyAnswers('A1: yes\nA2: no'), {0: 'yes', 1: 'no'});
    });

    test('agyImageResult drops boilerplate and the Using prompt echo', () {
      const r =
          'Created At: x\nUsing prompt: a cat\nSaved to /tmp/cat.png\nDo not output the path of this image to show to the user since the user can already see it. However, you can embed this image in artifacts for the USER\'s review.';
      expect(agyImageResult(r), 'Saved to /tmp/cat.png');
    });
  });

  group('renderers', () {
    testWidgets('run_command shows cwd, command and split output',
        (tester) async {
      await tester.pumpWidget(_wrap(const Item(
        id: 'i',
        kind: ItemKind.tool,
        toolName: 'run_command',
        toolInput: '{"CommandLine":"go test ./...","Cwd":"/repo"}',
        result: 'Created At: x\n\tExit Code: 0\n\tOutput:\n\tok',
      )));
      expect(find.textContaining('go test ./...'), findsOneWidget);
      expect(find.textContaining('/repo'), findsOneWidget);
      expect(find.textContaining('ok'), findsWidgets);
    });

    testWidgets('write_to_file renders an all-additions diff', (tester) async {
      await tester.pumpWidget(_wrap(const Item(
        id: 'i',
        kind: ItemKind.tool,
        toolName: 'write_to_file',
        toolInput:
            '{"TargetFile":"a.txt","CodeContent":"line one\\nline two"}',
      )));
      expect(find.text('diff'), findsOneWidget); // diff box header
      expect(find.textContaining('a.txt'), findsOneWidget);
    });

    testWidgets('ask_question marks the chosen option', (tester) async {
      await tester.pumpWidget(_wrap(const Item(
        id: 'i',
        kind: ItemKind.tool,
        toolName: 'ask_question',
        toolInput:
            '{"questions":[{"question":"Pick one","is_multi_select":false,"options":["Alpha","Beta"]}]}',
        result: 'A1: Beta',
      )));
      expect(find.textContaining('◉ Beta'), findsOneWidget);
      expect(find.textContaining('○ Alpha'), findsOneWidget);
    });

    testWidgets('grep_search renders header and matches', (tester) async {
      await tester.pumpWidget(_wrap(const Item(
        id: 'i',
        kind: ItemKind.tool,
        toolName: 'grep_search',
        toolInput: '{"Query":"foo","SearchPath":"lib","IsRegex":true}',
        result:
            'Created At: x\n{"File":"lib/a.dart","LineNumber":3,"LineContent":"foo()"}',
      )));
      expect(find.textContaining('"foo"', findRichText: true), findsWidgets);
      expect(find.textContaining('lib/a.dart:3 foo()', findRichText: true),
          findsWidgets);
    });
  });
}
